code/app/src/main/java/com/kernelcrew/moodapp/data/CommentProvider.java

```diff
@@ -0,0 +1,73 @@
+package com.kernelcrew.moodapp.data;
+
+import android.util.Log;
+
+import androidx.annotation.NonNull;
+
+import com.google.android.gms.tasks.Task;
+import com.google.firebase.auth.FirebaseAuth;
+import com.google.firebase.auth.FirebaseUser;
+import com.google.firebase.firestore.CollectionReference;
+import com.google.firebase.firestore.DocumentReference;
+import com.google.firebase.firestore.DocumentSnapshot;
+import com.google.firebase.firestore.EventListener;
+import com.google.firebase.firestore.FirebaseFirestore;
+import com.google.firebase.firestore.ListenerRegistration;
+import com.google.firebase.firestore.Query;
+import com.google.firebase.firestore.QuerySnapshot;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class CommentProvider {
+    private final CollectionReference collection;
+
+    private CommentProvider() {
+        FirebaseFirestore db = FirebaseFirestore.getInstance();
+        collection = db.collection("comments");
+    }
+
+    private static CommentProvider instance;
+
+    /**
+     * Get the singleton instance of the CommentProvider.
+     * @return Singleton instance
+     */
+    public static CommentProvider getInstance() {
+        if (instance == null) {
+            instance = new CommentProvider();
+        }
+
+        return instance;
+    }
+
+    public Task<Void> insertComment(Comment comment) {
+        DocumentReference docRef = collection.document();
+        comment.setId(docRef.getId());
+        return docRef.set(comment);
+    }
+
+    public Task<List<Comment>> getCommentsByMoodEventId(String moodEventId) {
+        return collection
+                .whereEqualTo("moodEventId", moodEventId)
+                .orderBy("created", Query.Direction.DESCENDING)
+                .get()
+                .continueWith(task -> {
```

Comment:
It would be cleaner (and more reliable) to use `onSuccessTask`. For example, see https://github.com/cmput301-w25/project-kernelcrew/blob/7d37c5b2de16890b5807dcde23df2c382809c1be/code/app/src/main/java/com/kernelcrew/moodapp/data/MoodEventProvider.java#L91-L98